# Ciclos definidos: _for_

Dentro de las estructuras de código, se encuentran aquellas que nos permiten realizar ciclos sobre alguna colección (_iterable_), recorriendo los elementos uno a uno y almacenándolo en una variable temporal. A estos ciclos se les conoce como **ciclos definidos**, ya que en principio no se debe cumplir ninguna condición para finalizar, simplemente acabar de iterar sobre los elementos de la colección. Para definir éste tipo de estrcutruas de flujo existe la partícula **for**.

## Estrcutrua de un ciclo definido

Sea una colección* 
```python
numeros_pares = [2, 4, 6, 8]
```

Entonces definimos el ciclio definido de la siguiente forma:
```python
for variable_local in numeros_pares:
    ··· codigo_a_ejecutar(variable_local)
```

Nótese que se declara una _variable_local_ en el ciclo, esta variable local es donde se irán alcmacenando los valores de la colección en cada ciclo.

*NOTA: CUANDO DECIMOS UNA COLECCIÓN, NOS REFERIMOS A CUALQUIER OBJETO QUE PUEDA ALMACENAR OBJETOS O QUE SIMPLEMENTE SEA UN ITERABLE. Ejemplo: `list`, `tuple`, `set`, `dict`, `range`, ...

In [2]:
numeros_pares = [2, 4, 6, 8]

In [5]:
for numero in numeros_pares:
    print(f"La variable local en este ciclo es {numero}") # Aquí la sentencia que se realizará en el ciclo es simplemente un print

La variable local en este ciclo es 2
La variable local en este ciclo es 4
La variable local en este ciclo es 6
La variable local en este ciclo es 8


El ciclo anterior iterará sobre todos los elementos de la colección, es decir, que realizará la setnencia `print(numero)` cuatro veces (o el número de veces correspondiente al tamaño de la colección) y el valor de la variable `numero` irá cambiando en cada iteración.



**NOTA: El nombre de la variable local (que siempre va entre el _for_ y el _in_) puede ser cualquiera, pero se recomienda, como con cualquier nombre de variable, que sea un nombre descriptivo, que tenga que ver con los elementos de la colección. Evitar el uso de variables como _i_, _x_, etc...** 

Un ciclo definido en C, se ve de la siguiente forma:
```c
for (i=0; i <= 100; i++)
{
   ··· codigo_a_ejecutar(i)
}
```
En C/C++ el ciclo no se realiza sobre un objeto iterable o colección, sino que se realiza literalmente sobre números que equivalen a índices de una lista.

## Iteración sobre listas

El ciclo definido _for_ se hace directo sobre el objeto iterable (o colección). En el caso de listas:

In [43]:
lista = ["manzana", "piña", "plátano"]

for fruta in lista:
    print(fruta)

manzana
piña
plátano


## Iteración sobre diccionarios

In [8]:
diccionario = {
    "México": "Ciudad de México",
    "Japón": "Tokio",
    "Argentina": "Buenos Aires"
}

In [10]:
# Este ciclo agarra como variable_local las llaves del diccionario
for variable in diccionario:
    print(variable)

México
Japón
Argentina


In [12]:
# Usando el método .values() en el diccionario, la variable local ahora agarra los valores
for variable in diccionario.values():
    print(variable)

Ciudad de México
Tokio
Buenos Aires


In [45]:
# Usan el método .items() en el diccionario, podemos hacer que se agarren tanto llaves como valores del dict
for variable in diccionario.items():
    print(variable) # La variable ahora es una tupla con las (llaves, valores)

('México', 'Ciudad de México')
('Japón', 'Tokio')
('Argentina', 'Buenos Aires')


In [51]:
# Pero podemos definir dos variables locales, usando las propiedades de tupla que da la "," en Python
for key, value in diccionario.items():
    #print(f"La llave es {key} y el valor es {value}")
    print(key, value)

México Ciudad de México
Japón Tokio
Argentina Buenos Aires


## Iteración sobre enumaraciones

Se puede usar la _built-in function_ `enumerate` para obtener, de una colección iterable, el índice que le toca a cada elemento.

In [22]:
paises = ["México", "Guatemala", "El Salvador", "Costa Rica"]

In [26]:
# El enumerate regresa una lista de tuplas con el índice y el elemento que le toca
for index, pais in enumerate(paises):
    print(index, pais)

0 México
1 Guatemala
2 El Salvador
3 Costa Rica


## Iteración usando range

Si queremos hacer iteración sobre números enteros podemos usar `range`

In [40]:
%%time
# range(<inicio>, <final>, <de cuanto en cuanto>)
for entero in range(1,10_000_000,1): # El final no lo toma en cuenta
    pass # La partícula pass basicamente no hace nada

CPU times: user 384 ms, sys: 0 ns, total: 384 ms
Wall time: 383 ms


In [41]:
# Nota: Para escribir números muy grandes en Python, se puede usar el guión bajo para separar de tres en tres como
# si fueran comas
100_000_000

100000000

## Estructuras de flujo anidadas

Es posible anidar estructuras de flujo, ya sean ciclos dentro de ciclos, condicionales dentro de ciclos o ciclcos dentro de condicionales, por mencionar algunas. Crear condicionales dentro de ciclos nos permite condicionarl el flujo según el valor que tenga la variable local

In [52]:
numeros_pares = [2, 4, 6, 8]

In [53]:
for numero in numeros_pares:
    if numero < 5:
        print(f"El número {numero} es menor que 5")
    else:
        print(f"El número {numero} es mayor o igual que 5")

El número 2 es menor que 5
El número 4 es menor que 5
El número 6 es mayor o igual que 5
El número 8 es mayor o igual que 5


### Ciclos anidados

Podemos realizar ciclos anidados sobre dos colecciones distintas de la siguiente forma:

In [74]:
ropa = ["pantalón", "saco"]
colores = ["azul", "negro", "beige", "gris"]

In [82]:
# Ejemplo de creación de listas con ciclos

# Vamos a crear un par de listas en las cuales se va a guardar todas las combinaciones de color
# para cada prenda
pantalones = [] # Se definen vacías para después, usando .append() poderlas rellenar
sacos = []

for prenda in ropa:
    for color in colores:
        if prenda == "pantalón":
            pantalones.append((prenda, color)) # Agrega la tupla (prenda, color)
        elif prenda == "saco":
            sacos.append((prenda, color))

In [87]:
for pantalon in pantalones:
    for saco in sacos:
        print(pantalon, saco)

('pantalón', 'azul') ('saco', 'azul')
('pantalón', 'azul') ('saco', 'negro')
('pantalón', 'azul') ('saco', 'beige')
('pantalón', 'azul') ('saco', 'gris')
('pantalón', 'negro') ('saco', 'azul')
('pantalón', 'negro') ('saco', 'negro')
('pantalón', 'negro') ('saco', 'beige')
('pantalón', 'negro') ('saco', 'gris')
('pantalón', 'beige') ('saco', 'azul')
('pantalón', 'beige') ('saco', 'negro')
('pantalón', 'beige') ('saco', 'beige')
('pantalón', 'beige') ('saco', 'gris')
('pantalón', 'gris') ('saco', 'azul')
('pantalón', 'gris') ('saco', 'negro')
('pantalón', 'gris') ('saco', 'beige')
('pantalón', 'gris') ('saco', 'gris')


In [84]:
sacos

[('saco', 'azul'), ('saco', 'negro'), ('saco', 'beige'), ('saco', 'gris')]

## Iteración sobre múltiples colecciones

Supongamos el caso donde tenemos dos colecciones que deseamos recorrer al mismo tiempo, como parejas ordenadas.

Ejemplo:

Tenemos dos listas del mismo tamaño, una con nombres y otra con apellidos:

In [60]:
nombres = ["Luis", "Julia", "Ana"]
apellidos_paternos = ["López", "González", "Ortiz"]
apellidos_maternos = ["Fernández", "Urrutia", "Pérez"]

Para hacer el recorrido en el _for_ sobre ambas listas al mismo tiempo, podemos usar la _built-in function_ `zip`.
Esta función lo que nos va a regresar es una lista de tuplas donde cada tupla contiene a los elementos de las listas

In [67]:
list(zip(nombres, apellidos_paternos))

[('Luis', 'López'), ('Julia', 'González'), ('Ana', 'Ortiz')]

In [64]:
for nombre, apellido_paterno, apellido_materno in zip(nombres, apellidos_paternos, apellidos_maternos):
    print(nombre, apellido_paterno, apellido_materno)

Luis López Fernández
Julia González Urrutia
Ana Ortiz Pérez


## Ejemplos de iteración

Durante una semana, Ana trabaja en un huerto de manzanas. Parte de su paga consiste en recibir una cierta cantidad de manzanas al final del día. Las manzanas que recibe cada día con 10. La pregunta es: ¿Cuántas manzanas tendrá al fin de la semana (trabajando 5 días)?

In [10]:
num_dias = 5
manzanas_ganadas = 10
canasta_manzanas = 0

for dia in range(1,num_dias+1):
    print(f"Al inicio del día {dia} la canasta tiene {canasta_manzanas} manzanas")
    manzanas_totales = canasta_manzanas + manzanas_ganadas
    canasta_manzanas = manzanas_totales
    print(f"Al final se tienen {canasta_manzanas} manzanas")
    print("#"*80)

Al inicio del día 1 la canasta tiene 0 manzanas
Al final se tienen 10 manzanas
################################################################################
Al inicio del día 2 la canasta tiene 10 manzanas
Al final se tienen 20 manzanas
################################################################################
Al inicio del día 3 la canasta tiene 20 manzanas
Al final se tienen 30 manzanas
################################################################################
Al inicio del día 4 la canasta tiene 30 manzanas
Al final se tienen 40 manzanas
################################################################################
Al inicio del día 5 la canasta tiene 40 manzanas
Al final se tienen 50 manzanas
################################################################################


In [11]:
num_dias = 5
manzanas_ganadas = 10
canasta_manzanas = 0

for dia in range(1,num_dias+1):
    print(f"Al inicio del día {dia} la canasta tiene {canasta_manzanas} manzanas")
    canasta_manzanas = canasta_manzanas + manzanas_ganadas
    print(f"Al final se tienen {canasta_manzanas} manzanas")
    print("#"*80)

Al inicio del día 1 la canasta tiene 0 manzanas
Al final se tienen 10 manzanas
################################################################################
Al inicio del día 2 la canasta tiene 10 manzanas
Al final se tienen 20 manzanas
################################################################################
Al inicio del día 3 la canasta tiene 20 manzanas
Al final se tienen 30 manzanas
################################################################################
Al inicio del día 4 la canasta tiene 30 manzanas
Al final se tienen 40 manzanas
################################################################################
Al inicio del día 5 la canasta tiene 40 manzanas
Al final se tienen 50 manzanas
################################################################################


In [12]:
num_dias = 5
manzanas_ganadas = 10
canasta_manzanas = 0

for dia in range(1,num_dias+1):
    print(f"Al inicio del día {dia} la canasta tiene {canasta_manzanas} manzanas")
    #canasta_manzanas = canasta_manzanas + manzanas_ganadas
    canasta_manzanas += manzanas_ganadas
    print(f"Al final se tienen {canasta_manzanas} manzanas")
    print("#"*80)

Al inicio del día 1 la canasta tiene 0 manzanas
Al final se tienen 10 manzanas
################################################################################
Al inicio del día 2 la canasta tiene 10 manzanas
Al final se tienen 20 manzanas
################################################################################
Al inicio del día 3 la canasta tiene 20 manzanas
Al final se tienen 30 manzanas
################################################################################
Al inicio del día 4 la canasta tiene 30 manzanas
Al final se tienen 40 manzanas
################################################################################
Al inicio del día 5 la canasta tiene 40 manzanas
Al final se tienen 50 manzanas
################################################################################


---

El problema consiste en calcular la inversión total despues de N años que se tiene al invertir una cantidad I de dinero, 
con una tasa de interés anual de T

In [15]:
años = 5
tasa = 20
inversion = 100

for año in range(1,años+1):
    print(f"La inversión inicial al año {año} es de {inversion} pesos")
    inversion_sig = inversion * (1 + tasa/100)
    inversion = inversion_sig
    print(f"La inversión al final del año {año} es de {inversion} pesos")
    print("#"*80)

La inversión inicial al año 1 es de 100 pesos
La inversión al final del año 1 es de 120.0 pesos
################################################################################
La inversión inicial al año 2 es de 120.0 pesos
La inversión al final del año 2 es de 144.0 pesos
################################################################################
La inversión inicial al año 3 es de 144.0 pesos
La inversión al final del año 3 es de 172.79999999999998 pesos
################################################################################
La inversión inicial al año 4 es de 172.79999999999998 pesos
La inversión al final del año 4 es de 207.35999999999999 pesos
################################################################################
La inversión inicial al año 5 es de 207.35999999999999 pesos
La inversión al final del año 5 es de 248.83199999999997 pesos
################################################################################


In [16]:
años = 5
tasa = 20
inversion = 100

for año in range(1,años+1):
    print(f"La inversión inicial al año {año} es de {inversion} pesos")
    inversion = inversion * (1 + tasa/100)
    print(f"La inversión al final del año {año} es de {inversion} pesos")
    print("#"*80)

La inversión inicial al año 1 es de 100 pesos
La inversión al final del año 1 es de 120.0 pesos
################################################################################
La inversión inicial al año 2 es de 120.0 pesos
La inversión al final del año 2 es de 144.0 pesos
################################################################################
La inversión inicial al año 3 es de 144.0 pesos
La inversión al final del año 3 es de 172.79999999999998 pesos
################################################################################
La inversión inicial al año 4 es de 172.79999999999998 pesos
La inversión al final del año 4 es de 207.35999999999999 pesos
################################################################################
La inversión inicial al año 5 es de 207.35999999999999 pesos
La inversión al final del año 5 es de 248.83199999999997 pesos
################################################################################


In [17]:
años = 5
tasa = 20
inversion = 100

for año in range(1,años+1):
    print(f"La inversión inicial al año {año} es de {inversion} pesos")
    inversion *= (1 + tasa/100)
    print(f"La inversión al final del año {año} es de {inversion} pesos")
    print("#"*80)

La inversión inicial al año 1 es de 100 pesos
La inversión al final del año 1 es de 120.0 pesos
################################################################################
La inversión inicial al año 2 es de 120.0 pesos
La inversión al final del año 2 es de 144.0 pesos
################################################################################
La inversión inicial al año 3 es de 144.0 pesos
La inversión al final del año 3 es de 172.79999999999998 pesos
################################################################################
La inversión inicial al año 4 es de 172.79999999999998 pesos
La inversión al final del año 4 es de 207.35999999999999 pesos
################################################################################
La inversión inicial al año 5 es de 207.35999999999999 pesos
La inversión al final del año 5 es de 248.83199999999997 pesos
################################################################################


---

Tenemos una lista de caracteres y queremos juntar en una sola palabra, todos y cada uno de los caracteres, usando el ciclo for

In [25]:
lista_letras = ["a","b","c","d","e"]

In [24]:
"".join(lista_palabras)

'abcde'

In [27]:
palabra = ""

for letra in lista_letras:
    palabra_final = palabra + letra 
    palabra = palabra_final

In [28]:
palabra

'abcde'

---

Calcula el factorial de 1000 utilizando el ciclo _for_. https://es.wikipedia.org/wiki/Factorial Para probar que salió bien, el factorial de 5 es 120.